Spike on assertContent test helper

ryanlabouve opened the pull request:
This PR adds  an `assertContent(assert, contentString, options)` test helper to allow assertion on the content of a tooltip.

This may be a basic implementation, but would love feedback on how to improve.

Also, I feel like I should add additional tests to cover passing in at least the selector options.

Is this heading in the right direction? 


(This relates to https://github.com/sir-dunxalot/ember-tooltips/issues/128)

sir-dunxalot commented on test-support/helpers/ember-tooltips.js:
```diff
@@ -139,3 +139,14 @@ export function assertTooltipSide(assert, options = {}) {
   }
 }
+
+export function assertContent(assert, contentString, options = {}) {
```
To match existing test helper conventions, we should name this helper `assertTooltipContent`.

sir-dunxalot replied:
Also, I wonder if `contentString` should be a part of `options`. For the spacing and side helpers I worked on, I passed all the options via the `options` block to standardize the params going to each helper.

sir-dunxalot commented on tests/integration/components/content-test.js:
```diff
@@ -0,0 +1,51 @@
+  this.render(hbs`
+    <div>
+      :)
+      {{tooltip-on-element text='Smiley face'}}
```
A good addition to this test will be a test for the block form:

```
<div>
  :)
  {{#tooltip-on-element}}
    Smiley face
  {{/tooltip-on-element}}
</div>
```

sir-dunxalot commented on README.md:
```diff
@@ -420,6 +420,7 @@ This addon exposes testing helpers which can be used inside of the consuming app
 * `assertTooltipRendered(assert)`: asserts if the tooltip has been rendered. When enableLazyRendering is true the tooltip will only be rendered after the user has interacted with the $target element. A tooltip can be rendered but not visible.
 * `assertTooltipNotRendered(assert)`: asserts if the tooltip has not been rendered. When enableLazyRendering is true the tooltip will only be rendered after the user has interacted with the $target element.
 * `assertTooltipSide(assert, { side: 'right' }): asserts that the tooltip is shown on the correct side of the target. Additional options that can be passed are `selector` and `targetSelector`.
+* `assertContent(assert, contentString): asserts that the content of the tooltip matches a given string. 
```
Seems good enough of a description for now. I'm going to be revamping the test helper documentation (#144) to give each test helpers its own heading and section in the readme.

sir-dunxalot commented:
Great timing! This looks like a great start and it may well be this simple.

One general improvement will be to use this test helper throughout the tooltip where we previously tested content with out. I found [tooltip-on-element-test](https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/tooltip-on-element-test.js), [tooltip-on-component-test(https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/tooltip-on-component-test.js), and [popover-on-component-test](https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/popover/popover-on-component-test.js) that should all be updated to use this new test helper. A quick search for `trim()` might reveal others.

ryanlabouve commented:
Cool! I made the changes to helper name and signature and added the block form helper.

> One general improvement will be to use this test helper throughout the tooltip where we previously tested content with out.

Are you wanting this to happen in this PR or in the future?

sir-dunxalot commented:
This PR please. It shouldn't be a big fix but will make sure our test suite is up-to-date with this new addition and that will also help to 'test' the test helper.

I believe this is the complete list of tests to update:
- [ ] [text-test](https://github.com/sir-dunxalot/ember-tooltips/blob/dad31a5e5fb04bf85fbc933efbfee3154b5a00d9/tests/integration/components/text-test.js#L18)
- [ ] [update-for-test](https://github.com/sir-dunxalot/ember-tooltips/blob/dad31a5e5fb04bf85fbc933efbfee3154b5a00d9/tests/integration/components/update-for-test.js#L44)
- [ ] [tooltip-on-element-test](https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/tooltip-on-element-test.js)
- [ ] [tooltip-on-component-test](https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/tooltip-on-component-test.js)
- [ ] [popover-on-component-test](https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/popover/popover-on-component-test.js)

sir-dunxalot commented:
Once those updates are done we are good to merge this and it will be a great and widely-used addition to this library!

ryanlabouve commented:
I added the tests to the spots you mentioned.

Here are a few questions / possible changes that I would like to make pending your 👍:
* allowing `targetSelector` to determine tooltip content via target (link to [source](https://github.com/sir-dunxalot/ember-tooltips/pull/146/files#diff-d2d31db5ee6b9cc51256e1f874f6acadR46))

* Show assertion if `contentString` is not passed in as an argument. This gives users of this test selector direct feedback if they are missing this required arg.

* Should we do something to enabled the test to work for lazyRendering? e.g. [here](https://github.com/sir-dunxalot/ember-tooltips/blob/master/tests/integration/components/tooltip-on-component-test.js#L33)

a15n commented:
> "Should we do something to enabled the test to work for lazyRendering?"

If a tooltip is not rendered then there will be no tooltip or tooltipText. I don't see how this test would work in that scenario. Am I understanding the question correctly?

ryanlabouve commented:
Perhaps I am misunderstanding, but lazy rendering is just waiting to stick the tooltips in the DOM until they are needed? 

If so, if I was doing an acceptance test I initially thought I would want to check their content regardless if they were rendered or not. But now that I'm thinking about it I guess that's a leaky abstraction.

a15n commented:
You're correct about how lazy rendering works.

There is no way to check their content unless they are rendered. So if you wanted to check their content in your acceptance test you'd need to do one of these things...

1. initialize `isShown=true` 
2. initialize with `enableLazyRendering=false`
3. `triggerTooltipEvent($tooltipTarget, 'mouseenter')`

sir-dunxalot commented:
> allowing targetSelector to determine tooltip content via target (link to source)

Could you expand on what capability you'd like to add?

> Show assertion if contentString is not passed in as an argument. This gives users of this test selector direct feedback if they are missing this required arg.

All for this! 👍  To be consistent with existing tests, `Ember.assert` could be used ([see example here](https://github.com/sir-dunxalot/ember-tooltips/blob/1e743ca332bc398869bc4729a8937fae811ec99e/addon/components/tether-tooltip-and-popover.js#L288).

sir-dunxalot commented on tests/integration/components/tooltip-on-element-test.js:
```diff
@@ -41,7 +42,11 @@ test('tooltip-on-element has the proper aria-describedby tag', function(assert)
 
-  assert.equal(this.$(`#${describedBy}`).text().trim(), 'Some info in a tooltip.');
+  assertTooltipContent(assert, {
+    // targetSelector: '.target',
```
The discussion about this line needs to be resolve int he comments.

sir-dunxalot commented:
One comment left to resolve (see prior comment) - otherwise the code changes look good!

sir-dunxalot commented:
@ryanlabouve Looks like the latest changes to master branch have created conflicts with this branch.

ryanlabouve commented:
Rebased off of master and fixed conflicts!

Also added the assertion we talked about and removed an unneeded comment.

Would you like me to fix the new ESLint errors as well? They are not related to this PR but it looks like they are causing the tests to fail.

sir-dunxalot commented:
Yes pleas, Ryan. Sorry for asking you to make unrelated changes. This is in part due to this PR being opened before a couple of bigger sweeping PRs were merged (addon linting being one of them).

ryanlabouve commented:
Cool. Just pushed the linting changes.

sir-dunxalot commented:
Great! I'll make a final review and should be able to approve changes before the end of this weekend.

This will be an awesome addition!

a15n commented:
good work Ryan

ryanlabouve commented:
Thanks!

a15n commented on README.md:
```diff
@@ -421,6 +421,7 @@ This addon exposes testing helpers which can be used inside of the consuming app
 * `assertTooltipNotRendered(assert)`: asserts if the tooltip has not been rendered. When enableLazyRendering is true the tooltip will only be rendered after the user has interacted with the $target element.
 * `assertTooltipSide(assert, { side: 'right' }): asserts that the tooltip is shown on the correct side of the target. Additional options that can be passed are `selector` and `targetSelector`.
 * `assertTooltipSpacing(assert, { side: 'right', spacing: 10 }): asserts that the tooltip is a given distance from the target (on a given side). `side` and `spacing` must be passed. Additional options that can be passed are `selector` and `targetSelector`.
+* `assertContent(assert, contentString): asserts that the content of the tooltip matches a given string. 
```
this needs to be updated...

```
assertTooltipContent(assert, { contentString: 'hello' }); .....
```

ryanlabouve replied:
Good catch!

a15n commented on test-support/helpers/ember-tooltips.js:
```diff
@@ -222,3 +222,20 @@ export function assertTooltipSpacing(assert, options) {
+export function assertTooltipContent(assert, options = {}) {
+  const { contentString } = options;
+
+  if (contentString === undefined) {
```
could you use http://emberjs.com/api/#method_isEmpty

ryanlabouve replied:
If we use `isEmpty` then a user would not be able to test the case where `contentString` is an empty string.

andrew-zenefits replied:
ah I meant 'isNone'

a15n commented on test-support/helpers/ember-tooltips.js:
```diff
@@ -222,3 +222,20 @@ export function assertTooltipSpacing(assert, options) {
+  const $tooltip = getTooltipFromBody(options.selector);
+  const tooltipContent = $tooltip.text().trim();
+
+  assert.equal(
```
nit: not blocking, could you follow the syntax we did elsewhere in this file...

```
assert.equal(a, b, 'some description');
```

PS: I like that you added the tooltipContent and contentString in the description

ryanlabouve replied:
👍

a15n commented on tests/integration/components/content-test.js:
```diff
@@ -0,0 +1,68 @@
+
+  assert.expect(2);
+
+  this.render(hbs`
```
the div is unnecessary. Could be written like...

```
this.render(hbs`{{tooltip-on-element test="foo"}}`);
```

ryanlabouve replied:
I think this makes the tests harder to read, but would be happy to change. My reasoning is that the surrounding div gives someone reading the test more context.

Again, happy to change, please just confirm.

andrew-zenefits replied:
@sir-dunxalot ?

sir-dunxalot replied:
To stay consistent with our other tests I think we should remove the `<div>` and use `this.render(hbs`{{tooltip-on-element test="foo"}}`);`.

ryanlabouve replied:
You got it. Updating now.

a15n commented on tests/integration/components/content-test.js:
```diff
@@ -0,0 +1,68 @@
+    contentString: 'Smiley face',
+  });
+
+  this.render(hbs`
```
could be simplified too (no div)

a15n commented on tests/integration/components/content-test.js:
```diff
@@ -0,0 +1,68 @@
+  assert.expect(2);
+
+  this.render(hbs`
+    <div>
```
everywhere

a15n commented on tests/integration/components/content-test.js:
```diff
@@ -0,0 +1,68 @@
+    </div>
+  `);
+
+  const stubbedAssert = {
```
can you explain what stubbedAssert does? I'm confused.

ryanlabouve replied:
This just stubs the `equal` on `assert.equal` that is passed in to our helper. The value this adds is testing that our helper calls what we think it will with the values it should have.

ryanlabouve commented:
Updates made! Please merge when you are able to confirm :-D

a15n commented:
LGTM

sir-dunxalot commented:
Merged! Release to come shortly.

sir-dunxalot commented:
Released as `2.8.0`. Thanks a lot @ryanlabouve!
